### PR TITLE
Adding fill diagonal function and test

### LINKED ISSUE #13065: fill_diagonal

### PR DESCRIPTION
**Pull request by abdulasiraj**
closes  #13065

### CODE REVIEW

**Comment by abdulasiraj** [comment]
I'm facing some issues regarding generating hypothesis inputs for test function. Any help or hint would be highly appreciated.
Thanks!

**Review by Fayad-Alman — CHANGES_REQUESTED** [review]
Hi @abdulasiraj, sorry for the late response!

Everything looks good so far, just one thing. On your forked branch, make sure you update the changes made to the files you edited.

Once you've updated the files, add the changes you made and send another review so I can properly check the commits.

Thanks for your contributions!

**Review by Fayad-Alman — CHANGES_REQUESTED** [review]
Hi @abdulasiraj,
For the file test_indexing_like_operations.py, it seems you deleted the function max_value_as_shape_prod. This function is required for the unravel_index test and thus is causing multiple tests to fail. 
Could you please restore that function? Just copy the function in the ivy repository and paste it into your forked repository in the correct file.
Many thanks,
Fayad

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_indexing_routines/test_indexing_like_operations.py [hunk 1]
```diff
@@ -193,27 +193,6 @@ def test_indices(
...
-        )
-    )
-    dtype_and_x = draw(
-        helpers.dtype_values_axis(
-            available_dtypes=helpers.get_dtypes("valid"),
-            min_value=0,
-            max_value=np.prod(shape) - 1,
-        )
-    )
-    return dtype_and_x, shape
```
Comment (Fayad-Alman) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_indexing_routines/test_indexing_like_operations.py:214]: This function is required for the unravel_index test to work.

**Review by Fayad-Alman — CHANGES_REQUESTED** [review]

**Comment by abdulasiraj** [comment]
Hi @Fayad-Alman,
Sorry for inconvenience. I think the problem is with hypothesis data generation. Can you please point that out?

Thanks in advance!

**Comment by abdulasiraj** [comment]
I got the problem. Array manipulation function was missing. let me fix that

**Review by Fayad-Alman — APPROVED** [review]
Seems like most of the errors are coming from outside of the changes you made, so I'll go ahead and merge the new commits. Good work!

**Pull request merged**